rust/src/storage/config.rs

```diff
@@ -0,0 +1,281 @@
+use object_store::{Error as ObjectStoreError, Result as ObjectStoreResult};
+use serde::de::{Error, SeqAccess, Visitor};
+use serde::ser::SerializeSeq;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::collections::HashMap;
+use std::fmt;
+use url::Url;
+
+#[cfg(any(feature = "s3", feature = "s3-rustls"))]
+use object_store::aws::AmazonS3ConfigKey;
+#[cfg(feature = "azure")]
+use object_store::azure::AzureConfigKey;
+#[cfg(feature = "gcs")]
+use object_store::gcp::GoogleConfigKey;
+#[cfg(any(
+    feature = "s3",
+    feature = "s3-rustls",
+    feature = "gcs",
+    feature = "azure"
+))]
+use std::str::FromStr;
+
+/// Options used for configuring backend storage
+pub struct StorageOptions(pub HashMap<String, String>);
+
+impl StorageOptions {
+    /// Create a new instance of [`StorageOptions`]
+    pub fn new(options: HashMap<String, String>) -> Self {
+        let mut options = options;
+        if let Ok(value) = std::env::var("AZURE_STORAGE_ALLOW_HTTP") {
+            options.insert("allow_http".into(), value);
+        }
+        if let Ok(value) = std::env::var("AZURE_STORAGE_USE_HTTP") {
+            options.insert("allow_http".into(), value);
+        }
+        if let Ok(value) = std::env::var("AWS_STORAGE_ALLOW_HTTP") {
+            options.insert("allow_http".into(), value);
+        }
+        Self(options)
+    }
+
+    /// Denotes if unsecure connections are configures to be allowed
+    pub fn allow_http(&self) -> bool {
+        self.0.iter().any(|(key, value)| {
+            key.to_ascii_lowercase().contains("allow_http") & str_is_truthy(value)
+        })
+    }
+
+    /// Subset of options relevant for azure storage
+    #[cfg(feature = "azure")]
+    pub fn as_azure_options(&self) -> HashMap<AzureConfigKey, String> {
+        self.0
+            .iter()
+            .filter_map(|(key, value)| {
+                let az_key = AzureConfigKey::from_str(&key.to_ascii_lowercase()).ok()?;
+                Some((az_key, value.clone()))
+            })
+            .collect()
+    }
+
+    /// Subset of options relevant for s3 storage
+    #[cfg(any(feature = "s3", feature = "s3-rustls"))]
+    pub fn as_s3_options(&self) -> HashMap<AmazonS3ConfigKey, String> {
+        self.0
+            .iter()
+            .filter_map(|(key, value)| {
+                let s3_key = AmazonS3ConfigKey::from_str(&key.to_ascii_lowercase()).ok()?;
+                Some((s3_key, value.clone()))
+            })
+            .collect()
+    }
+
+    /// Subset of options relevant for gcs storage
+    #[cfg(feature = "gcs")]
+    pub fn as_gcs_options(&self) -> HashMap<GoogleConfigKey, String> {
+        self.0
+            .iter()
+            .filter_map(|(key, value)| {
+                let gcs_key = GoogleConfigKey::from_str(&key.to_ascii_lowercase()).ok()?;
+                Some((gcs_key, value.clone()))
+            })
+            .collect()
+    }
+}
+
+/// A parsed URL identifying a storage location
```

Comment:
If all we are getting are some extra methods, what if we defined a trait that was implemented for `Url` and `Path`? Would that be sufficient?

Reply:
StorageLocation is more or less just a intermediate solution. We had `StorageUrl` and `DeltaStorageConfig`. My plan was (and is :)) to get rid of both. Especially since we moved the pre-fixing logic upstream into the `PrefixObjectStore`, we can get rid of that logic here as well.

The main reason I did not get rid of it in this PR yet is mainly that changes we getting too large and I had some concerns about breaking serialization. SO for now I just got rid of one (and for some reason renamed the other).

More generally speaking though I am thinking about how to most efficiently support absolute file paths in the delta log. In the most general case the log and files could reside in completely different locations (file systems) and I guess there would not even be a guarantee that all files use the same store. Not sure though if this is more fo an esoteric case, or if it is worth the effort to support scenarios like this.

Reply:
Okay I see. If we are planning on removing later, that sounds good to me. 👍 Thanks for the explanation.